Logging support

karmi opened the pull request:
This pull request adds support for logging the information about the request and response, common in other Elasticsearch clients.

To enable logging, users pass an `io.Writer` to the `LogOutput` field in the configuration:

```golang
elasticsearch.NewClient(elasticsearch.Config{
	LogOutput:       os.Stdout,
})
```

The client comes with four different log formatters:

* `LogFormatText` for plain text (default)
* `LogFormatColor` for a terminal-optimized, colored text
* `LogFormatCurl` for a runnable curl command
* `LogFormatJSON` for structured output

First three are mainly suitable for development, the JSON formatter is optimized for production use.

[Benchmark](https://github.com/elastic/go-elasticsearch/blob/logging/estransport/logger_benchmark_test.go) output:

```
BenchmarkTransport/Defaults         	 1000000	1192 ns/op	 736 B/op	10 allocs/op
BenchmarkTransportLogger/Text       	 1000000	2299 ns/op	 880 B/op	18 allocs/op
BenchmarkTransportLogger/Text-Body  	  300000	5550 ns/op	9394 B/op	33 allocs/op
BenchmarkTransportLogger/JSON       	 1000000	2137 ns/op	1184 B/op	13 allocs/op
BenchmarkTransportLogger/JSON-Body  	  500000	3727 ns/op	3872 B/op	22 allocs/op
```

The JSON formatter outputs an [ECS](https://github.com/elastic/ecs/#fields)-compatible structure.

See the [`_examples/logging.go`](https://github.com/elastic/go-elasticsearch/blob/logging/_examples/logging.go) for configuration examples and output.

NOTE: I've originally added support for a `LoggerFunc` in https://github.com/elastic/go-elasticsearch/commit/c75f586992fd5616e7e22acca4dc212e82c938b2, which would add an extension point for total control over logging, but decided to remove it in the end, because I felt it adds too much weight to the core code, and because the correct solution in the spirit of this package is to implement a custom transport.

axw commented on elasticsearch.go:
```diff
@@ -19,8 +20,16 @@ const (
+	// The HTTP transport object.
+	Transport http.RoundTripper
+
+	LogOutput       io.Writer             // Logging output
```
As neat as this is, I think it would be better to have a Logger interface, and allow the user to provide an implementation, like:

```go
logger := &elasticsearch.TextLogger{Output: os.Stdout}
// or &elasticsearch.JSONLogger{Output: os.Stdout, LogResponseBody: true}
// or &elasticsearch.CurlLogger{Output: os.Stdout, LogRequestBody: true}
client, err := elasticsearch.NewClient(elasticsearch.Config{
    Logger: logger,
})
```

That would make it more straightforward to integrate into popular loggers like logrus, zap, zerolog, etc. While it would be possible to do the same with a custom RoundTripper, doing it this way would take care of copying the body and such.

axw commented on estransport/estransport.go:
```diff
@@ -50,14 +62,36 @@ func New(cfg Config) *Client {
+	if c.logger != nil {
+		// TODO(karmi): Handle errors
+		// TODO(karmi): Handle closing
+		if req.Body != nil && req.Body != http.NoBody {
```
I led you astray here, you don't need all that complexity. In the APM agent we may not either; I wrote it that way so we could get the info up front, but that's probably unnecessary.

I think you can do it much more simply:

```go
func duplicateBody(req *http.Request) io.ReadCloser {
        if req.Body == nil || req.Body == http.NoBody {
                return http.NoBody
        }
        type readCloser struct {
                io.Reader
                io.Closer
        }
        var buf bytes.Buffer
        req.Body = &readCloser{Reader: io.TeeReader(req.Body, &buf), Closer: req.Body}
        return ioutil.NopCloser(&buf)
}
```

karmi commented:
We have discussed the pull request extensively with @axw over Zoom and Slack, and I've implemented all the requested changes in the closing commit 3fb2a27, which is a rebased and squashed combination of the WIP commits on the `logging` branch.
